components/sup/src/manager/service.rs

```diff
@@ -168,10 +169,110 @@ impl TemplateUpdate {
 enum InitializationState {
     Uninitialized,
     Initializing,
+    InitializerFailed,
     InitializerFinished,
     Initialized,
 }
 
+#[derive(Debug)]
+pub struct PersistentServiceWrapper {
+    run_state: ServiceRunState,
+    inner:     Option<Service>,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum RestartState {
+    None,
+    NeedsRestart,
+    Restarting,
+    Restarted,
+}
+
+impl Default for RestartState {
+    fn default() -> Self { RestartState::None }
+}
+
+#[derive(Debug, Default)]
+struct ServiceRunState {
+    ticks:            u64,
+    restart_attempts: u64,
+    restart_state:    RestartState,
+    restart_backoff:  Backoff,
+}
+
+impl PersistentServiceWrapper {
+    pub fn new(service: Service) -> PersistentServiceWrapper {
+        PersistentServiceWrapper { run_state: ServiceRunState::default(),
+                                   inner:     Some(service), }
+    }
+
+    pub fn service(&self) -> Option<&Service> { self.inner.as_ref() }
+
+    pub fn service_mut(&mut self) -> Option<&mut Service> { self.inner.as_mut() }
+
+    pub fn start(&mut self, service: Service) {
+        self.run_state.restart_state = match self.run_state.restart_state {
+            RestartState::None => RestartState::None,
+            RestartState::NeedsRestart => {
+                panic!("Start called on service which was not ready to be restarted");
```

Comment:
I think we want to make all of these `panic!` instances `error!` so that the text is emitted to stderr but the supervisor does not crash. We try to be super careful in the supervisor code to make sure that we do not panic.

Reply:
Any invalid usage of the internal restart state would cause a panic. If it did happen, it means we've essentially made a terrible mistake in our code. These code paths are essentially unreachable.